internal/image/image.go

```diff
@@ -73,6 +72,183 @@ func ScanImage(r reporter.Reporter, imagePath string) (ScanResults, error) {
 	return scannedLockfiles, err
 }
 
+type Image struct {
+	flattenedLayers []FileMap
+	innerImage      *v1.Image
+	extractDir      string
+}
+
+func (img *Image) LastLayer() *FileMap {
+	return &img.flattenedLayers[len(img.flattenedLayers)-1]
+}
+
+func (img *Image) Cleanup() error {
+	return os.RemoveAll(img.extractDir)
+}
+
+func loadImage(path string) (Image, error) {
+	image, err := tarball.ImageFromPath(path, nil)
+	if err != nil {
+		return Image{}, err
+	}
+
+	tempPath, err := os.MkdirTemp("", "osv-scanner-image-scanning-*")
+	if err != nil {
+		return Image{}, err
+	}
+
+	layers, err := image.Layers()
+	if err != nil {
+		return Image{}, err
+	}
+
+	outputImage := Image{
+		extractDir:      tempPath,
+		innerImage:      &image,
+		flattenedLayers: make([]FileMap, len(layers)),
+	}
+
+	// Reverse loop through the layers to start from the latest layer first
+	// this allows us to skip all files already seen
+	for i := len(layers) - 1; i >= 0; i-- {
+		hash, err := layers[i].DiffID()
+		hashStr := strings.TrimPrefix(hash.String(), "sha256:")
+		if err != nil {
+			return Image{}, err
+		}
+
+		dirPath := filepath.Join(tempPath, hashStr)
+		err = os.Mkdir(dirPath, 0755)
+		if err != nil {
+			return Image{}, err
+		}
+
+		layerReader, err := layers[i].Uncompressed()
+		if err != nil {
+			return Image{}, err
+		}
+		defer layerReader.Close()
+		tarReader := tar.NewReader(layerReader)
+
+		for {
+			header, err := tarReader.Next()
+			if errors.Is(err, io.EOF) {
+				break
+			}
+			if err != nil {
+				return Image{}, fmt.Errorf("reading tar: %w", err)
+			}
+			// Some tools prepend everything with "./", so if we don't Clean the
+			// name, we may have duplicate entries, which angers tar-split.
+			cleanedFileName := filepath.Clean(header.Name)
+			// Prevent "Zip Slip"
+			if strings.HasPrefix(cleanedFileName, "..") {
```

Comment:
"..foo" is a valid filename, so I think we need to add a filepath separator after ".." here.